### PR TITLE
Add `ConvertRawTxId`, use it for comparing `OneEraGenTxId`s

### LINKED ISSUE #1009: [BUG] - Inconsistency between MsgHasTx / MsgNextTx in the LocalTxMonitor
~~Internal~~/**External**

**Summary**

The LocalTxMonitor protocol doesn't _always_ respond appropriately to the `MsgHasTx` message. I've hunted down the issue and found that the root cause comes from the hard-fork combinator applied to the `GenTxId`. 

**Steps to reproduce**

1. Create a valid transaction
2. Submit the transaction through the local submit protocol
3. Before the transaction is removed from the mempool, acquire a mempool snapshot
4. Ask whether the transaction is in the mempool (`MsgHasTx`) using a TxId wrapped in any other era than the current one. 
5. Observe that the reply is `False` 
6. Ask the next transaction present in the mempool (`MsgNextTx`) to confirm that the transaction is indeed in the snapshot.

<details><summary>Here is a minimal reproduction example in Haskell (requires a serialized transaction to be passed as 1st argument)</summary>

```hs
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE QuantifiedConstraints #-}
module Main where

import Prelude

import qualified Codec.CBOR.Encoding as Cbor
import qualified Codec.CBOR.Write as Cbor
import Control.Monad.Class.MonadTimer
import Control.Tracer
import Data.ByteString
import Data.ByteString.Base16
import qualified Data.ByteString.Lazy as LBS
import Data.Map
import Data.Proxy
import qualified Data.Text as Text
import qualified Data.Text.Encoding as Text
import Data.Void
import System.Environment

import Cardano.Chain.Slotting
import Cardano.Ledger.Api
import Cardano.Ledger.Binary
import Cardano.Slotting.Slot

import Network.TypedProtocol
import Network.TypedProtocol.Codec

import Ouroboros.Consensus.Byron.Ledger.Config
import Ouroboros.Consensus.Cardano
import Ouroboros.Consensus.Cardano.Block
import Ouroboros.Consensus.Ledger.SupportsMempool
import Ouroboros.Consensus.Network.NodeToClient
import Ouroboros.Consensus.Node.NetworkProtocolVersion
import Ouroboros.Consensus.Shelley.Ledger
import Ouroboros.Consensus.Util

import Ouroboros.Network.Block
import Ouroboros.Network.Driver.Simple
import Ouroboros.Network.Magic
import Ouroboros.Network.Mux
import Ouroboros.Network.NodeToClient
import Ouroboros.Network.Protocol.LocalTxMonitor.Client
import qualified Ouroboros.Network.Protocol.LocalTxMonitor.Client as LTM
import Ouroboros.Network.Protocol.LocalTxSubmission.Client
import qualified Ouroboros.Network.Protocol.LocalTxSubmission.Client as LTS

import Ouroboros.Consensus.Protocol.Praos.Translate
    ()
import Ouroboros.Consensus.Shelley.Ledger.SupportsProtocol
    ()

type Block = CardanoBlock StandardCrypto

data NetworkName = Mainnet | PreProd | Preview deriving (Show, Read)

defaultNetwork :: NetworkName
defaultNetwork = Preview

defaultEpochSlots :: EpochSlots
defaultEpochSlots =
    case defaultNetwork of
        Mainnet -> EpochSlots 21600
        PreProd -> EpochSlots 21600
        Preview -> EpochSlots 4320

defaultNetworkMagic :: NetworkMagic
defaultNetworkMagic =
    NetworkMagic $ case defaultNetwork of
        Mainnet -> 764824073
        PreProd -> 1
        Preview -> 2

defaultNodeSocket :: FilePath
defaultNodeSocket = "/tmp/node.socket"

main :: IO ()
main = do
    transaction <- unsafeDecodeTransaction . unsafeDecodeBase16 . Prelude.head <$> getArgs
    let protocols = mkProtocols (mkLocalTxSubmissionClient transaction) (mkLocalTxMonitorClient transaction)
    withIOManager $ \iocp -> connectTo (localSnocket iocp) tracers (versions protocols) defaultNodeSocket
  where
    versions protocols =
        let
            v = NodeToClientV_15
            vData = NodeToClientVersionData defaultNetworkMagic False
         in
            simpleSingletonVersions v vData (protocols v)

    tracers =
        NetworkConnectTracers
            { nctMuxTracer = nullTracer
            , nctHandshakeTracer = nullTracer
            }

mkLocalTxSubmissionClient
    :: GenTx Block
    -> LocalTxSubmissionClient (GenTx Block) (HardForkApplyTxErr (CardanoEras StandardCrypto)) IO ()
mkLocalTxSubmissionClient transaction =
    LocalTxSubmissionClient $ pure $ SendMsgSubmitTx transaction $ \result -> do
        putStrLn $ "Submitted transaction " <> show (txId transaction) <> " -> " <> showResult result
        threadDelay maxBound
        pure $ LTS.SendMsgDone ()
  where
    showResult = \case
        SubmitSuccess -> "Success"
        SubmitFail reason -> "Failed: " <> show reason

mkLocalTxMonitorClient
    :: GenTx Block
    -> LocalTxMonitorClient (GenTxId Block) (GenTx Block) SlotNo IO ()
mkLocalTxMonitorClient transaction =
    LocalTxMonitorClient $ pure $ SendMsgAcquire $ \slot -> do
        putStrLn $ "Acquired mempool snapshot at slot: " <> show slot
        watchMempool []
  where
    watchMempool :: [GenTxId Block] -> IO (ClientStAcquired (GenTxId Block) (GenTx Block) SlotNo IO ())
    watchMempool !txs = pure $ SendMsgNextTx $ \case
        Just tx -> do
            watchMempool (txId tx : txs)
        Nothing -> do
            putStrLn $ "Mempool: " <> show txs
            pure $ if Prelude.null txs then
                SendMsgAwaitAcquire $ \slot -> do
                    putStrLn $ "Acquired mempool snapshot at slot: " <> show slot
                    watchMempool []
            else
                SendMsgHasTx (asTxIdBabbage transaction) $ \has -> do
                    putStrLn $ "Has " <> show (asTxIdBabbage transaction) <> "? " <> show has
                    pure $ SendMsgHasTx (asTxIdAlonzo transaction) $ \has' -> do
                        putStrLn $ "Has " <> show (asTxIdAlonzo transaction) <> "? " <> show has'
                        pure $ SendMsgRelease $ pure $ LTM.SendMsgDone ()

    asTxIdBabbage =
        txId

    asTxIdAlonzo =
        \case
            GenTxIdBabbage (ShelleyTxId i) -> GenTxIdAlonzo (ShelleyTxId i)
            orElse -> orElse
        .
        txId

unsafeDecodeBase16
    :: String
    -> ByteString
unsafeDecodeBase16 =
    either (error . show) id
    . decodeBase16Untyped
    . Text.encodeUtf8
    . Text.pack

unsafeDecodeTransaction
    :: ByteString
    -> GenTx Block
unsafeDecodeTransaction =
    either (error . show) GenTxBabbage
        . decodeFullDecoder'
            (eraProtVerLow @StandardBabbage)
            "Transaction"
            (fromPlainDecoder fromCBOR)
        . Cbor.toStrictByteString
        . wrapCBORinCBOR Cbor.encodePreEncoded

mkProtocols
    :: LocalTxSubmissionClient (GenTx Block) (HardForkApplyTxErr (CardanoEras StandardCrypto)) IO ()
    -> LocalTxMonitorClient (GenTxId Block) (GenTx Block) SlotNo IO ()
    -> NodeToClientVersion
    -> OuroborosApplicationWithMinimalCtx InitiatorMode addr LBS.ByteString IO () Void
mkProtocols localTxSubmissionClient localTxMonitorClient !v =
    NodeToClientProtocols
        { localTxSubmissionProtocol =
            protocol cTxSubmissionCodec (localTxSubmissionClientPeer localTxSubmissionClient)

        , localTxMonitorProtocol =
            protocol cTxMonitorCodec (localTxMonitorClientPeer localTxMonitorClient)

        , localChainSyncProtocol =
            protocol cChainSyncCodec chainSyncPeerNull

        , localStateQueryProtocol =
            protocol cStateQueryCodec localStateQueryPeerNull
        } `nodeToClientProtocols` v
  where
    protocol
        :: forall ps (st :: ps) pr failure bytes addr a.
            ( Show failure
            , forall (st' :: ps). Show (ClientHasAgency st')
            , forall (st' :: ps). Show (ServerHasAgency st')
            , ShowProxy ps
            )
        => (ClientCodecs Block IO -> Codec ps failure IO bytes)
        -> Peer ps pr st IO a
        -> RunMiniProtocolWithMinimalCtx InitiatorMode addr bytes IO a Void
    protocol codec peer =
        InitiatorProtocolOnly $ MiniProtocolCb $ \_ctx channel ->
            runPeer nullTracer (codec codecs) channel peer

    codecs :: ClientCodecs Block IO
    codecs = clientCodecs cfg (supportedVersions ! v) v
      where
        supportedVersions = supportedNodeToClientVersions (Proxy @Block)
        cfg = CardanoCodecConfig
            (let byron   = ByronCodecConfig defaultEpochSlots in byron)
            (let shelley = ShelleyCodecConfig in shelley)
            (let allegra = ShelleyCodecConfig in allegra)
            (let mary    = ShelleyCodecConfig in mary)
            (let alonzo  = ShelleyCodecConfig in alonzo)
            (let babbage = ShelleyCodecConfig in babbage)
            (let conway  = ShelleyCodecConfig in conway)
```
</details>

This is what the script above yields for a valid transaction (with some comments for clarity)

```
-- A mempool snapshot is acquired immediately. It is empty 
Acquired mempool snapshot at slot: SlotNo 44280567
Mempool: []

-- The transaction is successfully submitted
Submitted transaction HardForkGenTxId {getHardForkGenTxId = S (S (S (S (S (Z (WrapGenTxId {unwrapGenTxId = txid: TxId {unTxId = SafeHash "d7b4f21da71700e7fbbdea322c42f85b6373fecbbf16dab180b854df15503b66"}}))))))} -> Success

-- This causes a new mempool snapshot to become available. Which contains out transaction.
Acquired mempool snapshot at slot: SlotNo 44280567
Mempool: [HardForkGenTxId {getHardForkGenTxId = S (S (S (S (S (Z (WrapGenTxId {unwrapGenTxId = txid: TxId {unTxId = SafeHash "d7b4f21da71700e7fbbdea322c42f85b6373fecbbf16dab180b854df15503b66"}}))))))}]

-- Asking whether the mempool knows the tx id as a "BabbageGenTxId" returns `True`
Has HardForkGenTxId {getHardForkGenTxId = S (S (S (S (S (Z (WrapGenTxId {unwrapGenTxId = txid: TxId {unTxId = SafeHash "d7b4f21da71700e7fbbdea322c42f85b6373fecbbf16dab180b854df15503b66"}}))))))}? True

-- Asking whether the mempool knows *the same* tx id as a "AlonzoGenTxId" returns `False`
Has HardForkGenTxId {getHardForkGenTxId = S (S (S (S (Z (WrapGenTxId {unwrapGenTxId = txid: TxId {unTxId = SafeHash "d7b4f21da71700e7fbbdea322c42f85b6373fecbbf16dab180b854df15503b66"}})))))}? False
```

**Expected behavior**

When the transaction is in the mempool, `MsgHasTx` should return `True`, regardless of the era in which the `TxId` is being wrapped / submitted. We know it is in the mempool because it is returned in the next `MsgNextTx` call. 

I imagine that one might say: "But the transaction id isn't guaranteed to be equal and have compatible representation across eras!", yes. Sure. But in effect, it's been the case for ALL eras so far and given that the underlying type is the same and is era-independent, then the wrapping era we choose for `TxId` shouldn't matter. 

This is particularly annoying to deal with as it forces clients to _keep track_ of the current era also in the mempool monitoring protocol (this is already sufficiently annoying with the local-state-query 😬 ...). This is problematic especially around hard forks. 

Given that we look for the transaction id by looking at a membership in a set, a possible resolution path could be to derive the `Ord` instance (and while at it, the `Eq` too) on the inner type of the `TxId` type instance, instead of the type itself (i.e. `deriving newtype`). This way, it wouldn't matter where in the combinator is the underlying type, the equality/comparison will happen on the inner type which holds the era-independent transaction id.

   https://github.com/IntersectMBO/ouroboros-consensus/blob/c1b742690eb729c8bda0e103b76b8982584eaa5c/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HardFork/Combinator/Mempool.hs#L251-L254

**System info (please complete the following information):**
- OS Name: MacOS
- OS Version: 14.4
- Consensus version: 
    - any.ouroboros-consensus ==0.15.0.0,
    - any.ouroboros-consensus-cardano ==0.13.0.0,
    - any.ouroboros-consensus-diffusion ==0.10.0.0,
    - any.ouroboros-consensus-protocol ==0.7.0.0,
    
**Screenshots and attachments**
ø

**Additional context**
ø

### PR DESCRIPTION
**Pull request by amesgen**
Closes #1009

### CODE REVIEW

File: ouroboros-consensus-cardano/src/byron/Ouroboros/Consensus/Byron/Ledger/Mempool.hs [hunk 1]
```diff
@@ -162,6 +163,12 @@ instance HasTxId (GenTx ByronBlock) where
   txId (ByronUpdateProposal i _) = ByronUpdateProposalId i
   txId (ByronUpdateVote     i _) = ByronUpdateVoteId     i
 
+instance ConvertRawTxId (GenTx ByronBlock) where
+  toRawTxIdHash (ByronTxId             i) = CC.abstractHashToShort i
+  toRawTxIdHash (ByronDlgId            i) = CC.abstractHashToShort i
+  toRawTxIdHash (ByronUpdateProposalId i) = CC.abstractHashToShort i
+  toRawTxIdHash (ByronUpdateVoteId     i) = CC.abstractHashToShort i
```
Comment (amesgen) [inline ouroboros-consensus-cardano/src/byron/Ouroboros/Consensus/Byron/Ledger/Mempool.hs:170]: This instance assumes that it is not possible for two different constructors to contain the same hash. This seems reasonable, as the payload of these different kinds of transactions are completely different. However, to be extra safe, we could eg prepend a tag byte.

File: ouroboros-consensus-diffusion/changelog.d/20240325_124036_alexander.esgen_compare_txid_on_raw_hash.md
Comment (jasagredo) [inline ouroboros-consensus-diffusion/changelog.d/20240325_124036_alexander.esgen_compare_txid_on_raw_hash.md:1]: Not necessary. You are modifying tests not the library

**Review by jasagredo — APPROVED** [review]
Rest looks good

File: ouroboros-consensus-diffusion/changelog.d/20240325_124036_alexander.esgen_compare_txid_on_raw_hash.md
Comment (amesgen) [reply ouroboros-consensus-diffusion/changelog.d/20240325_124036_alexander.esgen_compare_txid_on_raw_hash.md:1]: Oh right, will replace it with an empty file :+1:

File: ouroboros-consensus-cardano/src/byron/Ouroboros/Consensus/Byron/Ledger/Mempool.hs (hunk 1 above)
Comment (amesgen) [reply ouroboros-consensus-cardano/src/byron/Ouroboros/Consensus/Byron/Ledger/Mempool.hs:170]: This seems fine, especially as Byron transactions aren't being newly submitted anymore.

**Pull request merged**